### PR TITLE
Add `split` command

### PR DESCRIPTION
**Pull request by Hellysonrp**
Add `split` command, to load each chapter of a YouTube video as a song in the queue.

I'm creating this as a draft to gather feedback, and also because I have to review the code and improve some things.

Example:
Screenshot_15

One doubt that I have is about this piece of code:
https://github.com/codetheweb/muse/blob/dcac22832d02376c351603c944aabec17f2b7d1b/src/services/player.ts#L374-L377

I've removed the `+7`, because for me this was just skipping 7 seconds of the music.
I'm not sure if this should be removed though. If it's there, it's there for some reason, right?
I'm using the Windows version of FFmpeg.

### CODE REVIEW

**Comment by codetheweb** [comment]
Nice idea!

My main comment would be that if you think split should be a separate command (rather than an argument to `.play`) the common logic between the split and play commands needs to be extracted out somewhere (possibly a service that takes a URL and a few options?).

-------

> I've removed the +7, because for me this was just skipping 7 seconds of the music.
I'm not sure if this should be removed though. If it's there, it's there for some reason, right?
I'm using the Windows version of FFmpeg.

This does not skip 7 seconds of music for me, at least for the ffmpeg version I'm using on macOS (4.4) and the version of ffmpeg inside the built image (codetheweb/muse). Tested by skipping to 3:50 in a song that was 3:54 and audio still played for 3-4 seconds.

It's possible that the Windows version of ffmpeg acts slightly differently for some reason. Are you able to try it from within the built container image?

**Comment by Hellysonrp** [comment]
> My main comment would be that if you think split should be a separate command (rather than an argument to .play) the common logic between the split and play commands needs to be extracted out somewhere (possibly a service that takes a URL and a few options?).

I thought of making it an argument to the `play` command, but I couldn't think of a way to parse the arguments. For example, if someone wants to use the `i` (or `immediate`) argument with a `split` argument, how this could be achieved?
I agree that the common logic must be extracted to somewhere else to be used by both the commands.

> It's possible that the Windows version of ffmpeg acts slightly differently for some reason. Are you able to try it from within the built container image?

I'll try it soon.

**Comment by Hellysonrp** [comment]
I just tried to run [this video](https://www.youtube.com/watch?v=JoTnfmJ-PqA), both in my "production" bot (running with the latest docker image in a dedicated server) and in my modified code (tried in a docker image and in Windows), and tried to seek to `1:35:35`. In my "production" bot, 7 seconds of the video are skipped; the other bot runs the music fine, with no extra or missing seconds.

> Tested by skipping to 3:50 in a song that was 3:54 and audio still played for 3-4 seconds.

I believe that it worked for you because of the cache:
https://github.com/codetheweb/muse/blob/fe1717788f28e847cdb23f7361bd68bb1302fb4a/src/services/player.ts#L361-L363

The video that I'm using for testing is more than an hour and a half long, so it'll never be cached.
If a video is cached, the parameter is passed as-is, without adding 7 seconds:
https://github.com/codetheweb/muse/blob/fe1717788f28e847cdb23f7361bd68bb1302fb4a/src/services/player.ts#L315-L321

In conclusion, I'm confident that the `+7` can be safely removed.

About the code improvements, I'll get into it this weekend.

**Comment by codetheweb** [comment]
Yeah, it looks like the cache was impacting my ability to reproduce that. I see now what you mean.

However removing that seek throws off soundbite commands (i.e. `packers` or `bears`) and changing the seek position of those soundbites doesn't seem to help. I assume there's some subtle race condition somewhere but I don't currently have the time to debug it. If you can get those working I'm happy to remove the 7s seek.

**Comment by codetheweb** [comment]
Hey @Hellysonrp were you able to get this working?

**Comment by Hellysonrp** [comment]
I haven't had time to work on it yet.
I've been busy with IRL stuff.
Hopefully I'll be able to get to work on it soon.

**Comment by codetheweb** [comment]
No rush.

**Comment by Hellysonrp** [comment]
I finally had time to rebase this code.
Most of the code is copy and paste from the code I had before, adapting to fit the new command.
Let me know if something is wrong or could be changed.

**Comment by codetheweb** [comment]
Thanks for finishing this!

I did a bit of cleanup in [610ed42](https://github.com/codetheweb/muse/pull/363/commits/610ed42cfc94f796925c03b4e668efd864acd613), please see if that looks good.

I also changed the title format slightly to put the chapter name first. When the video title is long it can cut off the chapter title:

**Before**:
Screen Shot 2022-03-07 at 10 11 40 PM

**After**:
Screen Shot 2022-03-07 at 10 12 01 PM

**Comment by Hellysonrp** [comment]
https://github.com/codetheweb/muse/blob/610ed42cfc94f796925c03b4e668efd864acd613/src/services/get-songs.ts#L297

Every video description is parsed here, even if `shouldSplitChapters` is `false`.
I don't know the performance impact this could cause. I'm worried about big playlists. One of my friends has a playlist with more than a thousand videos and we often listen to that playlist.
Other than that, everything looks good to me. If the performance impact is negligible, then I'm just overthinking (again 😅).

**Comment by codetheweb** [comment]
Good point. I suspect the performance impact would be negligible but I'm too lazy to test it so I just changed it in e126047 😛.

**Comment by codetheweb** [comment]
Released in v1.3.0.
